src/control/probe.rs

```diff
@@ -0,0 +1,463 @@
+use crate::util::{likely, unlikely};
+use core::{marker::PhantomData, ptr::NonNull, slice};
+
+use super::{BitMaskIter, Group, Tag};
+
+/// Probe over an individual group, returned by [`Probe`].
+#[derive(Clone)]
+pub(crate) struct GroupProbe {
+    /// Original control pointer.
+    ctrl: NonNull<Tag>,
+
+    /// Original search tag.
+    hash_tag: Tag,
+
+    /// Group being probed.
+    group: Group,
+
+    /// Index of the group.
+    index: usize,
+
+    /// Number of buckets, minus one.
+    bucket_mask: usize,
+}
+impl GroupProbe {
+    /// Checks whether the group contains any empty slots.
+    ///
+    /// This is equivalent to `empty().is_some()`.
+    #[inline]
+    pub(crate) fn has_empty(&self) -> bool {
```

Comment:
This was used in `RawTableInner::find_or_find_insert_slot_inner`, `RawTableInner::find_inner`, and `RawIterHashInner::next`.